hazelcast-client/src/test/java/classloading/ThreadLeakClientTest.java

```diff
@@ -42,13 +51,81 @@ public void testThreadLeak() {
-        clientConfig.getNetworkConfig()
-                .setConnectionAttemptLimit(Integer.MAX_VALUE);
+        clientConfig.getConnectionStrategyConfig().setAsyncStart(true);
+        clientConfig.getNetworkConfig().setConnectionAttemptLimit(Integer.MAX_VALUE);
 
         HazelcastInstance client = HazelcastClient.newHazelcastClient(clientConfig);
 
         client.shutdown();
     }
+
+    @Test
+    public void testThreadLeakWhenClientCanNotStart() {
+        try {
+            HazelcastClient.newHazelcastClient();
+            Assert.fail();
+        } catch (IllegalStateException e) {
+        }
+    }
+
+    @Test
+    public void testThreadLeakWhenClientCanNotStartDueToAuthenticationError() {
+        try {
+            Hazelcast.newHazelcastInstance();
+            ClientConfig config = new ClientConfig();
+            config.getGroupConfig().setName("invalid cluster");
+            HazelcastClient.newHazelcastClient(config);
+            Assert.fail();
+        } catch (IllegalStateException e) {
+        } finally {
+            Hazelcast.shutdownAll();
+        }
+    }
+
+    @Test
+    public void testThreadLeakWhenClientCanNotConstructDueToNoMemberDiscoveryStrategyConfig() {
+        try {
+            ClientConfig config = new ClientConfig();
+            config.getNetworkConfig().getDiscoveryConfig().addDiscoveryStrategyConfig(
+                    new DiscoveryStrategyConfig(new ClientDiscoverySpiTest.NoMemberDiscoveryStrategyFactory(),
+                            Collections.<String, Comparable>emptyMap()));
+            HazelcastClient.newHazelcastClient(config);
+            Assert.fail();
+        } catch (IllegalStateException e) {
+        }
+    }
+
+    @Test
+    public void testThreadLeakWhenClientCanNotStartDueToIncorrectUserCodeDeploymentClass() {
+        try {
+            Hazelcast.newHazelcastInstance();
+
+            ClientConfig config = new ClientConfig();
+            ClientUserCodeDeploymentConfig clientUserCodeDeploymentConfig = new ClientUserCodeDeploymentConfig();
+            clientUserCodeDeploymentConfig.addClass("invalid.class.test");
+            config.setUserCodeDeploymentConfig(clientUserCodeDeploymentConfig.setEnabled(true));
+
+            HazelcastClient.newHazelcastClient(config);
+            Assert.fail();
+        } catch (HazelcastException e) {
+        } finally {
+            Hazelcast.shutdownAll();
+        }
+    }
+
+
+    @Test
+    public void testThreadLeakWhenClientCanNotConstructDueToIncorrectSerializationServiceFactoryClassName() {
+        try {
+            ClientConfig config = new ClientConfig();
+            SerializationConfig serializationConfig = new SerializationConfig();
+            serializationConfig.addDataSerializableFactoryClass(5, "invalid.factory");
+            config.setSerializationConfig(serializationConfig);
+
+            HazelcastClient.newHazelcastClient(config);
+            Assert.fail();
+        } catch (HazelcastSerializationException e) {
```

Comment:
Can't we remove this catch and replace it with 

@Test(expected=HazelcastSerializationException). 

Same goes for the other catches.

Reply:
done.